app/controllers/UkimsNumberController.scala

```diff
@@ -47,27 +54,71 @@ class UkimsNumberController @Inject() (
-    implicit request =>
+  def onPageLoadCreate(mode: Mode): Action[AnyContent] =
+    (identify andThen checkProfile andThen getData andThen requireData) { implicit request =>
       val preparedForm = request.userAnswers.get(UkimsNumberPage) match {
         case None        => form
         case Some(value) => form.fill(value)
       }
 
-      Ok(view(preparedForm, mode))
-  }
+      Ok(view(preparedForm, routes.UkimsNumberController.onSubmitCreate(mode)))
+    }
 
-  def onSubmit(mode: Mode): Action[AnyContent] = (identify andThen getData andThen requireData).async {
+  def onSubmitCreate(mode: Mode): Action[AnyContent] = (identify andThen getData andThen requireData).async {
     implicit request =>
       form
         .bindFromRequest()
         .fold(
-          formWithErrors => Future.successful(BadRequest(view(formWithErrors, mode))),
+          formWithErrors =>
+            Future.successful(BadRequest(view(formWithErrors, routes.UkimsNumberController.onSubmitCreate(mode)))),
           value =>
             for {
               updatedAnswers <- Future.fromTry(request.userAnswers.set(UkimsNumberPage, value))
               _              <- sessionRepository.set(updatedAnswers)
             } yield Redirect(navigator.nextPage(UkimsNumberPage, mode, updatedAnswers))
         )
   }
+
+  def onPageLoadUpdate: Action[AnyContent] =
+    (identify andThen getData andThen requireData) { implicit request =>
+      val preparedForm = request.userAnswers.get(UkimsNumberUpdatePage) match {
+        case None        => form
+        case Some(value) => form.fill(value)
+      }
+
+      Ok(view(preparedForm, routes.UkimsNumberController.onSubmitUpdate))
+    }
+
+  def onSubmitUpdate: Action[AnyContent] = (identify andThen getData andThen requireData).async { implicit request =>
+    form
+      .bindFromRequest()
+      .fold(
+        formWithErrors =>
+          Future.successful(BadRequest(view(formWithErrors, routes.UkimsNumberController.onSubmitUpdate))),
+        value =>
+          request.userAnswers.set(UkimsNumberUpdatePage, value) match {
+            case Success(answers) =>
+              sessionRepository.set(answers).flatMap { _ =>
+                UpdateTraderProfile.buildUkimsNumber(answers, request.eori) match {
+                  case Right(model) =>
+                    traderProfileConnector.updateTraderProfile(model, request.eori).map { _ =>
+                      Redirect(navigator.nextPage(UkimsNumberUpdatePage, NormalMode, answers))
+                    }
+                  case Left(errors) => Future.successful(logErrorsAndContinue(errors))
+                }
+              }
+          }
+      )
+  }
+
+  def logErrorsAndContinue(errors: data.NonEmptyChain[ValidationError]): Result = {
+    val errorMessages = errors.toChain.toList.map(_.message).mkString(", ")
+
+    // TODO: got to profile page
+    val continueUrl = RedirectUrl(routes.HomePageController.onPageLoad().url)
```

Comment:
Why do we send to homepage if something goes wrong? I'd expect journery recovery, or is there a reason for this being the home page? There seems to be a todo also

Reply:
i'm waiting for Yohan's profile page - hold your horses :P

Reply:
holding :)

Reply:
holding :)